### PR TITLE
GH-37574: [Python] Compatibility with numpy 2.0

### LINKED ISSUE #37574: [Python] Compatibility with numpy 2.0 (removal of np.unicode_)
### Describe the bug, including details regarding any error messages, version, and platform.

Numpy has deprecated `np.unicode_` and removed it in an upcoming 2.0 release. Current nightly builds do not have `np.unicode_` which results in an incompatibility between pyarrow main and numpy main due to this mapping:

https://github.com/apache/arrow/blob/13f0cd8ee44833efbced05c8505de12f9f02f235/python/pyarrow/pandas_compat.py#L87-L102

and possibly would also fail on this one

https://github.com/apache/arrow/blob/13f0cd8ee44833efbced05c8505de12f9f02f235/python/pyarrow/pandas_compat.py#L1008-L1018

numpy now raises `AttributeError: `np.unicode_` was removed in the NumPy 2.0 release. Use `np.str_` instead.` See the [full traceback](https://github.com/geopandas/geopandas/actions/runs/6086861805/job/16514196822?pr=2996#step:5:4662) in the geopandas CI.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by graingert**
### What changes are included in this PR?

support for numpy 2
install numpy 2 and pandas 2.2.0.dev0 from scientific-python-nightly-wheels

### Are these changes tested?
I tested this locally with numpy==2.0.0.dev0 and pandas==2.2.0.dev0+325.g6c58a217f5

### Are there any user-facing changes?
No

* Closes: #37574

### CODE REVIEW

File: ci/scripts/install_pandas.sh [hunk 1]
```diff
@@ -28,7 +28,7 @@ pandas=$1
 numpy=${2:-"latest"}
 
 if [ "${numpy}" = "nightly" ]; then
-  pip install --extra-index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple --pre numpy
```
Comment (graingert) [inline ci/scripts/install_pandas.sh:31]: scipy-wheels-nightly only has up to [pandas-2.1.0.dev0+926](https://pypi.anaconda.org/scipy-wheels-nightly/simple/pandas/2.1.0.dev0%2B926.g4e8306690e/pandas-2.1.0.dev0%2B926.g4e8306690e-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl)

scientific-python-nightly-wheels has up to
[pandas-2.2.0.dev0+325](https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/pandas/2.2.0.dev0%2B325.g6c58a217f5/pandas-2.2.0.dev0%2B325.g6c58a217f5-cp39-cp39-macosx_11_0_arm64.whl)

File: python/pyarrow/pandas_compat.py [hunk 2]
```diff
@@ -36,6 +36,17 @@
...
 
+try:
+    _np_unicode = np.unicode_
+except AttributeError:
+    _np_unicode = np.str_
+
+try:
+    _np_sctypes = np.sctypes
+except AttributeError:
+    from numpy.core.numerictypes import sctypes as _np_sctypes
```
Comment (AlenkaF) [inline python/pyarrow/pandas_compat.py:47]: Thank you for contributing, much appreciated!

One question, is the reason for the Try Except older Python version (or numpy?) or can we just replace `np.unicode_` with `np.str_` and `np.sctypes` with `sctypes` (like `np.NaN` with `np.nan`)?

Comment (graingert) [reply python/pyarrow/pandas_compat.py:47]: I'm just trying out a run locally with numpy 1.16.6

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit test-conda-python-3.10-pandas-nightly

**Comment by graingert** [comment]
@jorisvandenbossche the nightly build passed!

**Review by AlenkaF — APPROVED** [review]
LGTM, thanks!

**Review by jorisvandenbossche — APPROVED** [review]
Thanks a lot!

**Pull request merged**